File: coding/src/com/algos/li13_graph/revised/ShortestPath_ShortestPathInUndirectedGraph.java
package  com.algos.li13_graph.revised;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;

import  com.algos.li30_model.GraphAdjLst;

public class ShortestPath_ShortestPathInUndirectedGraph {

	public static void main(String[] args) {
		ShortestPath_ShortestPathInUndirectedGraph obj = new ShortestPath_ShortestPathInUndirectedGraph();
		GraphAdjLst g = new GraphAdjLst();
		g.addEdge(0, 1);
		g.addEdge(0, 2);
		g.addEdge(1, 3);
		g.addEdge(3, 0);
		g.addEdge(3, 4);
		
		System.out.println(obj.shortedPath(g, 0, 4));

		int[] dist = obj.shortedPath(g.adj, 5, 0);
		System.out.println(Arrays.toString(dist));
	}

	// Idea: Do bfs and insert lengths also in queue.
	private int shortedPath(GraphAdjLst g, int src, int dest) {
		int n = g.adj.size();

		Queue<int[]> queue = new LinkedList<>();
		int[] vis = new int[n];
		queue.add(new int[]{src, 0});
		vis[src] = 1;
		while(!queue.isEmpty()) {
			int[] curr = queue.remove();
			if(curr[0] == dest) 
				return curr[1];
			
			for(int neigh: g.adj.get(curr[0])) {
				if(vis[neigh] == 0) {
					vis[neigh] = 1;
					queue.add(new int[]{neigh, curr[1]+1});
				}
			}
		}
		return -1;
	}

	// Problem: https://practice.geeksforgeeks.org/problems/shortest-path-in-undirected-graph-having-unit-distance/
	private int[] shortedPath(HashMap<Integer, ArrayList<Integer>> adj, int n, int src) {
		Queue<Integer> q = new LinkedList<>();
		int[] dist = new int[n];
		Arrays.fill(dist, Integer.MAX_VALUE);
		
		dist[0] = 0;
		q.add(src);
		while(!q.isEmpty()) {
			int node = q.remove();
			for(int neigh: adj.get(node)) {
				if(dist[node]+1 < dist[neigh]) {
					dist[neigh] = dist[node]+1;
					q.add(neigh);
				}
			}
		}

		return dist;
	}

}

File: dsa/concepts/13_graphs/python/shortest_path_shortest_path_in_undirected_graph.py
from collections import deque
import math


"""
Idea:
Since the graph is undirected and all edges have unit weight,
the shortest path can be found using BFS.

Approach:
1. Use BFS starting from source.
2. Each BFS level represents distance +1.
3. First time reaching a node guarantees the shortest distance.

Time Complexity: O(V + E)
Space Complexity: O(V)
"""
def shortest_path_to_dest(adj, src, dest):
    n = len(adj)
    visited = [False] * n
    q = deque()

    q.append((src, 0))
    visited[src] = True

    while q:
        node, dist = q.popleft()
        if node == dest:
            return dist

        for neigh in adj[node]:
            if not visited[neigh]:
                visited[neigh] = True
                q.append((neigh, dist + 1))

    return -1


"""
Problem:
https://practice.geeksforgeeks.org/problems/shortest-path-in-undirected-graph-having-unit-distance/

Idea:
Compute shortest distance from source to ALL nodes using BFS.
Distance relaxations are always +1 since unit edges.

Time Complexity: O(V + E)
Space Complexity: O(V)
"""
def shortest_path_all(adj, src):
    n = len(adj)
    dist = [math.inf] * n
    q = deque()

    dist[src] = 0
    q.append(src)

    while q:
        node = q.popleft()
        for neigh in adj[node]:
            if dist[node] + 1 < dist[neigh]:
                dist[neigh] = dist[node] + 1
                q.append(neigh)

    return dist


# one test (same as Java main)
if __name__ == "__main__":
    adj = [[] for _ in range(5)]

    adj[0].append(1)
    adj[0].append(2)
    adj[1].append(3)
    adj[3].append(0)
    adj[3].append(4)

    print(shortest_path_to_dest(adj, 0, 4))
    print(shortest_path_all(adj, 0))
